### PR TITLE
Add `fakeTimersLolex` to implement JestEnvironment interface

### PR DESCRIPTION
**Pull request by douglascayers**
Opening pull request per [this conversation](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/bbe8520b9e8121f63b30facb7a152695e226ddfd#r38589515).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - https://github.com/facebook/jest/releases/tag/v25.1.0
  - https://github.com/facebook/jest/pull/8925
  - https://github.com/facebook/jest/commit/0935f7c2cd7780e47365511e35efc829ba359df4#diff-e8e69a5b9fa4b43b2e24abe1933d2855R44
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

### CODE REVIEW

**Comment by douglascayers** [comment]
I've reviewed the build errors, but I don't know how to resolve them.

[Azure Pipelines](https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=42003&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=c264d665-0b79-59b7-7e04-9965cbe74860&l=188) build result:

```
types/jest/index.d.ts(486,51): error TS2307: Cannot find module 'jest-diff'.
types/jest/index.d.ts(540,44): error TS2307: Cannot find module 'pretty-format'.

Error: Compilation had errors
    at process.<anonymous> (/home/vsts/work/1/s/node_modules/@definitelytyped/perf/dist/measure/measureBatchCompilationWorker.js:24:19)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
Error: Parsing failed.
    at fail (/home/vsts/work/1/s/node_modules/@definitelytyped/utils/dist/process.js:76:20)
    at ChildProcess.<anonymous> (/home/vsts/work/1/s/node_modules/@definitelytyped/utils/dist/process.js:63:21)
    at ChildProcess.emit (events.js:311:20)
    at finish (internal/child_process.js:861:14)
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
```

[Travis CI](https://travis-ci.org/github/DefinitelyTyped/DefinitelyTyped/builds/677028976#L392) build result:

```
Error: Errors in typescript@3.9 for external dependencies:
../jest/index.d.ts(486,51): error TS2307: Cannot find module 'jest-diff' or its corresponding type declarations.
../jest/index.d.ts(540,44): error TS2307: Cannot find module 'pretty-format' or its corresponding type declarations.
    at /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:195:19
    at Generator.next (<anonymous>)
    at fulfilled (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:6:58)
Error: The following packages had errors: jest-environment-puppeteer
    at doRunTests (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/bin/tester/test-runner.js:220:11)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by douglascayers** [comment]
The compile errors are in a separate package, `jest`, and not related to my changes (as far as I can tell).

If you open https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts locally, you should see two compile errors:

1. [line 486](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L486) at `import("jest-diff").DiffOptions`

2. [line 540](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L540) at `import('pretty-format').Plugin`

What is the recommended next action considering the issue is due to a package I'm not trying to update??

**Review by favna — APPROVED** [review]
LGTM. As for the failing CI, I *think* that including the dependencies in the package.json list of dependencies will resolve the issue but I'm not 100% sure on that. You can definitely try and run tests locally (`npm run test`)

**Comment by douglascayers** [comment]
> LGTM. As for the failing CI, I _think_ that including the dependencies in the package.json list of dependencies will resolve the issue but I'm not 100% sure on that. You can definitely try and run tests locally (`npm run test`)

Unfortunately, updating the `package.json` file didn't resolve the build issues.

image

image

**Comment by douglascayers** [comment]
> LGTM. As for the failing CI, I _think_ that including the dependencies in the package.json list of dependencies will resolve the issue but I'm not 100% sure on that. You can definitely try and run tests locally (`npm run test`)

After my [first attempt](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44033#issuecomment-617496471) didn't work, I recalled that [jest-diff](https://github.com/facebook/jest/tree/master/packages/jest-diff) and [pretty-format](https://github.com/facebook/jest/tree/master/packages/pretty-format) are sub-packages of the [jest monorepo](https://github.com/facebook/jest).

So I then tried again by adding the `@jest` prefix to the package names:

image

This led to an interesting different error after running `npm run test`:

```
Error: In package.json: Dependency @jest/jest-diff not in whitelist.
If you are depending on another `@types` package, do *not* add it to a `package.json`. Path mapping should make the import work.
For namespaced dependencies you then have to add a `paths` mapping from `@namespace/library` to `namespace__library` in `tsconfig.json`.
If this is an external library that provides typings,  please make a pull request to types-publisher adding it to `dependenciesWhitelist.txt`.
    at checkPackageJsonDependencies (/Users/doug.ayers/Documents/workspace/DefinitelyTyped/node_modules/types-publisher/bin/lib/definition-parser.js:187:19)
```

Thinking that `jest-diff` and `pretty-format` were part of DefinitelyTyped, I started down the route of adding a `paths` mapping to `tsconfig.json` according to the first part of the error message:
> If you are depending on another `@types` package, do *not* add it to a `package.json`. Path mapping should make the import work.
For namespaced dependencies you then have to add a `paths` mapping from `@namespace/library` to `namespace__library` in `tsconfig.json`.

Realizing that those packages are in fact **not** part of DefinitelyTyped, then they must be external. And the next part of the error message had guidance on external dependencies:
> If this is an external library that provides typings,  please make a pull request to types-publisher adding it to `dependenciesWhitelist.txt`.

Taking a peek at [dependenciesWhitelist.txt](https://github.com/microsoft/types-publisher/blob/master/dependenciesWhitelist.txt), I saw that some other `@jest` packages were already listed.

I then verified that `jest-diff` and `pretty-format` are indeed written in typescript and provide their own typings -- meaning they're not candidates to bring in to DefinitelyTyped.

In conclusion, I've submitted https://github.com/microsoft/types-publisher/pull/769 to add `@jest/jest-diff` and `@jest/pretty-format` to the dependencies whitelist.

**Comment by jablko** [comment]
Is this a bug in `@jest/environment`?

```Shell
$ npm install @jest/environment
$ > index.ts
$ tsc --types @jest/environment index.ts
node_modules/@jest/environment/build/index.d.ts:8:23 - error TS2688: Cannot find type definition file for 'jest'.

8 /// <reference types="jest" />
                        ~~~~
```

`@jest/environment` references `types="jest"` without a dependency on `@types/jest`.

I don't know why `@jest/environment/build/index.d.ts` contains `/// <reference types="jest" />`: I suspect it's unintended. I haven't found an existing issue for this yet.

**Comment by jablko** [comment]
(You get the `jest-diff` and `pretty-format` errors in the automated tests because in that case *all* `@types` packages exist but `@types/jest`'s dependencies aren't installed because `@jest/environment` references `types="jest"` without a dependency on `@types/jest`?)

**Comment by jablko** [comment]
I opened facebook/jest#9875 to remove the `/// <reference types="jest" />` from `@jest/environment`. If it's accepted then this PR should start passing the automated tests, because `@types/jest-environment-puppeteer` won't depend on `jest-diff`/`pretty-format` (via `@jest/environment` and `@types/jest`).

**Comment by douglascayers** [comment]
Thanks @jablko for the help!

**Comment by douglascayers** [comment]
Update: @jablko's pull request https://github.com/facebook/jest/pull/9875 got merged into `master` branch of facebook/jest.

I imagine once they cut a new [release](https://github.com/facebook/jest/releases) (25.4.1+) then we can retry testing this pull request (I think..)

**Comment by Favna** [comment]
@DouglasAntunes Tip: once that's released, other than waiting for staff to retrigger the CI processes, since PRs on this repo are squash and merged anyway you can just trigger it yourself at that time with an empty commit, in CLI: `git commit --allow-empty -m "trigger ci" && git push`

**Comment by Favna** [comment]
@sandersn this PR is still pending work elsewhere and should not be closed yet. Can you prevent @typescript-bot from doing so?

**Comment by douglascayers** [comment]
That's unfortunate that @typescript-bot closed this in-progress pull request

**Comment by Favna** [comment]
@DouglasAntunes I guess just rebase your branch when Jest is updated and recreate the PR *sigh*

**Comment by orta** [comment]
oops, sorry about the bot - we added support today to make sure it won't close if maintainers are still chatting https://github.com/DefinitelyTyped/dt-mergebot/pull/33
